src/spectral.ts

```diff
@@ -25,9 +28,16 @@ export class Spectral {
     this.resolver = opts && opts.resolver ? opts.resolver : new Resolver();
   }
 
-  public async run(target: object): Promise<IRuleResult[]> {
-    const resolvedTarget = (await this.resolver.resolve(target)).result;
-    return runRules(target, this.rules, this.functions, { resolvedTarget });
+  public async run(target: IParserMeta | object | string): Promise<IRuleResult[]> {
```

Comment:
Instead of messing with the target argument, could we possibly add a second optional argument for this stuff? 

then we can throw warnings saying "We cannot give you any positional information because you didnt give us enough information" and the output can be ?? or something? Does that make any sense?

Reply:
Yeah, we had that done this way initially as discussed here https://github.com/stoplightio/spectral/pull/106#discussion_r268676451

Reply:
Ah, my comment saying `Yeah having it there but hidden from docs is fine with me.` was written from mobile and I hadn't seen marcs suggestion.

Ok, so the name IParserMeta is whats throwing me. Marcs example makes it clear that it can either be a parsed AST object, or a generic object, and if its the generic object then its not gonna work with line numbers and stuff. That's all fine if we use a name a bit more like his. 

Metadata is usually a different thing to data, so if it has meta in the name I'd expect it to be a different argument.

Reply:
@philsturgeon 
Yeah, naming is certainly not my strongest asset :laughing:

I can name it `IParsedAst` if it makes everything more clear.

Reply:
Renamed to `IParsedResult` since it also has some different properties than just `ast`, so this could be confusing.
Let me know what you think @philsturgeon

Reply:
> and if its the generic object then its not gonna work with line numbers and stuff

It should still work if go with original suggestion. Basically, if it's a plain object we stringily it and parse it with our parser to get positional info:

```ts
run(target: IParserAstResult | object | string) {
  let parsed = target;
  if (!isParserAstResult(target)) {
    parsed = parseWithPointers(safeStringify(target));
  }

  // ...resolve parsed, etc
}
```

Reply:
Yeah, we reparse it therefore some ranges are attached. Initially, when we didn't reparse the object, ranges were missing.

Reply:
Yeah, we reparse it therefore some ranges are attached. Initially, when we didn't reparse the object, ranges were missing.